compiler-rt/lib/gwp_asan/platform_specific/guarded_pool_allocator_posix.cpp

```diff
@@ -72,12 +72,15 @@ void GuardedPoolAllocator::unreserveGuardedPool() {
         State.GuardedPagePoolEnd - State.GuardedPagePool);
 }
 
-void GuardedPoolAllocator::allocateInGuardedPool(void *Ptr, size_t Size) const {
+bool GuardedPoolAllocator::allocateInGuardedPool(void *Ptr, size_t Size) const {
   assert((reinterpret_cast<uintptr_t>(Ptr) % State.PageSize) == 0);
   assert((Size % State.PageSize) == 0);
-  Check(mprotect(Ptr, Size, PROT_READ | PROT_WRITE) == 0,
-        "Failed to allocate in guarded pool allocator memory");
-  MaybeSetMappingName(Ptr, Size, kGwpAsanAliveSlotName);
+  // mprotect might fail because of system limitation. we don't expect a die
```

Comment:
Looking at [the documented return values / errors of mprotect](https://man7.org/linux/man-pages/man2/mprotect.2.html), there can be errors
- because the programmer provided mprotect() bad arguments (e.g. EINVAL, EACCES) or
- because of reasons the programmer has no influence over (e.g. ENOMEM).

While before, every error was treated overly pessimistically (I assume that `Check` kills the process in Debug builds and probably also in Release builds), we now treat errors overly optimistically. For example, errors of the first category are ignored which doesn't sound very appealing to me.

Question: Do you know more specifically, which type of error this fix addresses? If yes, it would be nice to fix the problem more fine-granularly.

Reply:
Some stress tests fail at this point but unfortunately, I don't know why. See https://play.clickhouse.com/play?user=play#c2VsZWN0CnRlc3RfY29udGV4dF9yYXcsCnB1bGxfcmVxdWVzdF9udW1iZXIgYXMgcHIsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMgYXMgc3RhdHVzLCByZXBvcnRfdXJsCmZyb20gY2hlY2tzIHdoZXJlIGNoZWNrX3N0YXJ0X3RpbWUgPj0gbm93KCkgLSBJTlRFUlZBTCAyNCBIT1VSIGFuZCB0ZXN0X2NvbnRleHRfcmF3IT0nJwphbmQgcHVsbF9yZXF1ZXN0X251bWJlcj0wIGFuZCBjaGVja19uYW1lIGlsaWtlICclc3RyZXNzJScgYW5kICh0ZXN0X25hbWUgaWxpa2UgJyVmYXRhbCUnIG9yIHRlc3RfbmFtZSBpbGlrZSAnJXNpZ25hbCUnKSBhbmQgdGVzdF9jb250ZXh0X3JhdyBsaWtlICclZ3dwJScKb3JkZXIgYnkgY2hlY2tfc3RhcnRfdGltZSBkZXNj

Reply:
But this fix is safe because we simply go aground gwp-asan as long as it gets mprotect failure.
tcmalloc just does the same thing without checking errno.https://github.com/google/tcmalloc/blob/5034f8cecdbe559bf24e0ae7f7eb7c10b873ac9e/tcmalloc/guarded_page_allocator.cc#L94

but they think failure when dealloc pointer is unacceptable, so I didn't change other logic

Reply:
Okay, I am fine with the fix then and I will approve.

Just a few minor things:
- "we don't expect a die" --> "we don't expect to die".
- in 3rd-party (contrib) code, it is nice to wrap custom patches in comments, e.g.

```cpp
/// Start Clickhouse-specific code
...
/// End ClickHouse-specific code
```

Perhaps that is something you could add?

Btw., to find out due to which error the stress tests failed, you could have made a test PR (marked "wip"/not for merging) where an exception with the actual error code is thrown. These exceptions should (theoretically) surface in the CI stress tests. This will of course only work if the failures occur not too sporadically.

Reply:
Good suggestion. Although the failure occurs very occasionally, it's always good to write a test PR

Reply:
splitting VMA will increase the map size by at most 2, so it should not be a bug problem. But actually I don't know why CI failed, and right now only fails in debug mode.

Reply:
>and right now only fails in debug mode

Likely because in debug mode `MMAP_THRESHOLD` is 16K, while in release - 64MiB - https://github.com/ClickHouse/ClickHouse/blob/23281c73482ab9285bf6f592488b2d683dd8abb3/src/Common/Allocator.cpp#L20